display: Add support for an ST7789V based LCD

marc-cpdesign opened the pull request:
This adds a driver for st7789v lcd controller, and TL019FQV01 lcd.
The bulk of the driver is based on the existing ili9340 driver.

Signed-off-by: Marc Reilly <marc@cpdesign.com.au>

marc-cpdesign committed:
display: Add support for an ST7789V based LCD

This adds a driver for st7789v lcd controller, and TL019FQV01 lcd.
The bulk of the driver is based on the existing ili9340 driver.

Signed-off-by: Marc Reilly <marc@cpdesign.com.au>

vanwinkeljan reviewed (approved):
Looks good to me.

Any chance to also add a sample so that in case of driver changes the sample can be used?

marc-cpdesign committed:
samples: display: add a simple sample for st7789v display

This adds a basic app which puts some rectangles on the display

It's hard to write an equivalent README to the rest of the sample
documentation, as the LCD I have uses a custom board thrown together for
testing. (ie other LCDs have publicly available modules)

Signed-off-by: Marc Reilly <marc@cpdesign.com.au>

marc-cpdesign commented:
Thanks for reviewing.!

> Any chance to also add a sample so that in case of driver changes the sample can be used?

I added a small sample, FWIW
I also updated some of the GPIO --> GPIOS in the defines

dbkinder commented on samples/display/st7789v/README.rst:
```diff
@@ -0,0 +1,41 @@
+| P0.02       | RESET       |
++-------------+-------------+
+
+You might need to alter these according to your specific board/LCD configuration.
```
Can you add the actual example commands to build/flash/run the sample, using the ``.. zephyr-app-commands`` directive. (See other sample docs for an example.

marc-cpdesign replied:
Added.

dbkinder commented on samples/display/st7789v/README.rst:
```diff
@@ -0,0 +1,41 @@
+
+Overview
+********
+This sample will draw some basic rectangles onto the display.
```
Should make a reference to the hardware, assuming this is the right link:
```
This sample draws rectangles onto an `ST7789V display
<https://www.crystalfontz.com/controllers/Sitronix/ST7789V/>.
```
Or is there a link to the specific display module you tested this with?

marc-cpdesign replied:
I added a couple of links, but as I say in the comments, the specific module that I dev/tested this with was a custom board.

dbkinder reviewed (changes requested):
Looks good, but needs sample build commands.

vanwinkeljan commented on samples/display/st7789v/sample.yaml:
```diff
@@ -0,0 +1,7 @@
+  name: st7789v_sample
+tests:
+  sample.display.st7789v:
+    build_only: true
```
Missing `platform_whitelist: nrf52_pca10040`

vanwinkeljan reviewed (changes requested):
Thx for the sample on small remark on the sample.yaml (should solve Shippable failures).
Also add a license header to the main.c

dbkinder reviewed (approved):
doc changes LGTM, thanks.

MaureenHelm commented on drivers/display/Kconfig.st7789v:
```diff
@@ -0,0 +1,68 @@
+	bool "ST7789V display driver"
+	depends on SPI
+	help
+	Enable driver for ST7789V display driver.
```
Please indent the help text by two spaces

marc-cpdesign replied:
done

MaureenHelm commented on drivers/display/display_st7789v.c:
```diff
@@ -0,0 +1,368 @@
+#include <spi.h>
+#include <gpio.h>
+#include <misc/byteorder.h>
+#include <display.h>
```
Please update the include paths to reflect changes in #16540

marc-cpdesign replied:
done

MaureenHelm replied:
spi, gpio, and display paths also need to be updated

MaureenHelm commented on samples/display/st7789v/boards/nrf52_pca10040.conf:
```diff
@@ -0,0 +1,6 @@
+
+CONFIG_ST7789V=y
+CONFIG_ST7789V_LCD_TL019FQV01=y
+CONFIG_ST7789V_RGB565=y
```
Nothing in this file appears to be specific to the nrf board. Is there a reason you didn't just put it all in the `proj.conf` file?

marc-cpdesign replied:
move these to proj.conf

marc-cpdesign commented:
Updated for Maureen's requested changes, rebased to current master

zzzh commented:
Hello,
I have a ST7789V based LCD module ( a 1.54" one, not TL019FQV01), and nrf52_pca10040 board,  so I did a try on this sample.
Unfortunately,  I got "display_st7789v: Could not get SPI device for LCD", I checked the code and realized that it can't find it from the inited driver.
I don't understand why, any ideas?

vanwinkeljan commented:
@zzzh could you try to use SPI 1 instead of SPI 0 by changing it in `samples/display/st7789v/nrf52_pca10040.overlay`?

https://github.com/zephyrproject-rtos/zephyr/blob/b631cc09abb4a660b824bffee1e3aba0ab6c7b46/samples/display/st7789v/nrf52_pca10040.overlay#L6-L8

zzzh commented:
@vanwinkeljan Thanks for replying.
I did not try that, and don't think that's the way to solve it.
Actually, I'd fixed it by adding a line in proj.conf : CONFIG_SPI_x=y, x is the number of SPI channel you chose in the overlay file.
